drivers/bicker_ser.c

```diff
@@ -0,0 +1,751 @@
+
+#define TOUINT(ch)	((unsigned)(uint8_t)(ch))
+#define BYTESWAP(in)	((((uint16_t)(in) & 0x00FF) << 8) + (((uint16_t)(in) & 0xFF00) >> 8))
+
+#ifdef WORDS_BIGENDIAN
+#define SWAPONBE(in)	BYTESWAP(in)
+#else
+#define SWAPONBE(in)	(in)
+#endif
+
+upsdrv_info_t upsdrv_info = {
+	DRIVER_NAME,
+	DRIVER_VERSION,
+	"Nicola Fontana <ntd@entidi.it>",
+	DRV_EXPERIMENTAL,
+	{ NULL }
+};
+
+typedef struct {
+	uint8_t  id;
+	uint16_t min;
+	uint16_t max;
+	uint16_t std;
+	uint8_t  enabled;
+	uint16_t val;
+} BickerParameter;
+
+/**
+ * Send a packet.
+ * @param idx     Command index
+ * @param cmd     Command
+ * @param data    Source data or NULL for no data field
+ * @param datalen Size of the source data field or 0
+ * @return        `datalen` on success or -1 on errors.
+ */
+static ssize_t bicker_send(char idx, char cmd, const void *data, size_t datalen)
+{
+	char buf[BICKER_PACKET];
+	size_t buflen;
+	ssize_t ret;
+
+	ser_flush_io(upsfd);
+
+	if (data != NULL) {
+		if (datalen > BICKER_MAXDATA) {
+			upslogx(LOG_ERR, "Data size exceeded: %d > %d",
+				(int)datalen, BICKER_MAXDATA);
+			return -1;
+		}
+		memcpy(buf + 4, data, datalen);
+	} else {
+		datalen = 0;
+	}
+
+	buf[0] = BICKER_SOH;
+	buf[1] = datalen + BICKER_HEADER;
+	buf[2] = idx;
+	buf[3] = cmd;
+	buf[1 + BICKER_HEADER + datalen] = BICKER_EOT;
+	buflen = 1 + BICKER_HEADER + datalen + 1;
+
+	ret = ser_send_buf(upsfd, buf, buflen);
+	if (ret < 0) {
+		upslog_with_errno(LOG_WARNING, "ser_send_buf failed");
+		return -1;
+	} else if ((size_t) ret != buflen) {
+		upslogx(LOG_WARNING, "ser_send_buf returned %d instead of %d",
+			(int)ret, (int)buflen);
+		return -1;
+	}
+
+	upsdebug_hex(3, "bicker_send", buf, buflen);
+	return datalen;
+}
+
+/**
+ * Receive a packet with a data field of unknown size.
+ * @param idx  Command index
+ * @param cmd  Command
+ * @param data Destination buffer or NULL to discard the data field
+ * @return     The size of the data field on success or -1 on errors.
+ *
+ * The data field is stored directly in the destination buffer. `data`,
+ * if not NULL, must have at least BICKER_MAXDATA bytes.
+ */
+static ssize_t bicker_receive(char idx, char cmd, void *data)
+{
+	ssize_t ret;
+	size_t datalen;
+	char buf[BICKER_PACKET];
+
+	/* Read first two bytes (SOH + size) */
+	ret = ser_get_buf_len(upsfd, buf, 2, BICKER_TIMEOUT, 0);
+	if (ret < 0) {
+		upslog_with_errno(LOG_WARNING, "Initial ser_get_buf_len failed");
+		return -1;
+	} else if (ret < 2) {
+		upslogx(LOG_WARNING, "Timeout waiting for response packet");
+		return -1;
+	} else if (buf[0] != BICKER_SOH) {
+		upslogx(LOG_WARNING, "Received 0x%02X instead of SOH (0x%02X)",
+			TOUINT(buf[0]), TOUINT(BICKER_SOH));
+		return -1;
+	}
+
+	datalen = buf[1] - BICKER_HEADER;
+
+	/* Read the rest: command index (1 byte), command (1 byte), data
+	 * (datalen bytes) and EOT (1 byte), i.e. `datalen + 3` bytes */
+	ret = ser_get_buf_len(upsfd, buf + 2, datalen + 3, BICKER_TIMEOUT, 0);
```

Comment:
Here 3 would ideally be `1 + BICKER_HEADER + 1  -  2` because it is the packet length - datalen - 2 (bytes already read). I admit that is confusing as hell.

Basically I read the packet in two steps: (1) the initial 2 bytes (where the size field resides) and (2) the rest of the packet. I tried to clarify the process in commit c78524c3f99b50ce3c08eedafb5bb0ac2eb0411e: let me know if that helps.